pkg/algorithm/nodeselect/v1alpha1/select_node.go

```diff
@@ -187,6 +210,96 @@ func (ac *Config) getBlockDevice() (*ndmapis.BlockDeviceList, error) {
+func (ac *Config) ClaimBlockDevice(nodeBDs *nodeBlockDevice, spc *apis.StoragePoolClaim) (*NodeClaimedBDDetails, error) {
+	nodeClaimedBDs := &NodeClaimedBDDetails{
+		NodeName:        "",
+		BlockDeviceList: []ClaimedBDDetails{},
+	}
+
+	namespace := env.Get(env.OpenEBSNamespace)
+	bdcKubeclient := bdc.NewKubeClient().
+		WithNamespace(namespace)
+	labels := map[string]string{string(apis.StoragePoolClaimCPK): spc.Name}
+	nodeClaimedBDs.NodeName = nodeBDs.NodeName
+
+	for _, bdName := range nodeBDs.BlockDevices.Items {
+		var hostName, bdcName string
+		claimedBD := ClaimedBDDetails{}
+
+		bdObj, err := ac.BlockDeviceClient.Get(bdName, metav1.GetOptions{})
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to get block device {%s}", bdName)
+		}
+		hostName = bdObj.Labels[string(apis.HostNameCPK)]
+
+		if bdObj.IsClaimed() {
+			//TODO: Use HasLabel Predicate for below check
+			bdcName = bdObj.Spec.ClaimRef.Name
+			bdcObj, err := bdcKubeclient.Get(bdcName, metav1.GetOptions{})
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to get block device claim{%s}", bdcName)
+			}
+			if bdcObj.Labels[string(apis.StoragePoolClaimCPK)] != spc.Name {
+				return nil, errors.Errorf("block device %s is already in use", bdName)
+			}
+		} else {
+			bdcName = "bdc-" + string(bdObj.UID)
+			capacity := volume.ByteCount(bdObj.Spec.Capacity.Storage)
+			//TODO: Move below code to some function
+			bdcObj, err := bdc.NewBuilder().
+				WithName(bdcName).
+				WithNamespace(namespace).
+				WithLabels(labels).
+				WithBlockDeviceName(bdName).
+				WithHostName(hostName).
+				WithCapacity(capacity).Build()
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to build blockd device claim for bd {%s}", bdName)
+			}
+
+			_, err = bdcKubeclient.Create(bdcObj.Object)
+			if err != nil {
+				return nil, errors.Wrapf(err, "failed to create block device claim for bdc {%s}", bdcName)
+			}
+
+			isClaimed, err := ac.waitForClaimedStatus(bdName)
+			if err != nil {
+				//TODO: Handle these things from review suggesstions
+				_ = bdcKubeclient.Delete(bdcName, &metav1.DeleteOptions{})
+				return nil, errors.Wrapf(err, "failed to claim block device bd {%s}", bdName)
+			}
+			if !isClaimed {
+				_ = bdcKubeclient.Delete(bdcName, &metav1.DeleteOptions{})
+				return nil, errors.Wrapf(err, "failed to claim block device: %s", bdName)
+			}
+		}
+		claimedBD.DeviceID = bdObj.GetDeviceID()
+		claimedBD.BDName = bdObj.Name
+		nodeClaimedBDs.BlockDeviceList = append(nodeClaimedBDs.BlockDeviceList, claimedBD)
+	}
+	return nodeClaimedBDs, nil
+}
+
+// waitForClaimStatus return the claim status of block device
+func (ac *Config) waitForClaimedStatus(bdName string) (bool, error) {
+	for i := 0; i < retryCount; i++ {
+		bdObj, err := ac.BlockDeviceClient.Get(bdName, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		if bdObj.IsClaimed() {
+			return true, nil
+		}
+		time.Sleep(waitTime)
+	}
+	return false, nil
+}
```

Comment:
Will need to find some other way -- I do not prefer code having `sleeps` 
Let us offload this to reconciliation.

Reply:
Overall -- 
SPC application should just lead to following :
1. Do I need to create pools ? If yes -- proceed for pool creation

Steps for pool creation :
1. Get me BDs from the SPC that satisfies following:
 i) BD is not used in other pool provisioning
 ii) BD is not claimed and if it is claimed it has the label of the current SPC
2. Once you get the BD  list -- there will be some BDs that will be unclaimed. if all are calimed go to step 4.
3. Pass the unclaimed BDs to a function that creates claim for that and do not wait for claim to be successful.

4. After above -- call algorithm that gives you a nodeBDs structure. This algorithm should only give you BDs that is claimed and have the current SPC as label and is active.
5. Rest -- as usual, the process.

Reply:
If the claim is not successful do not wait and pool creation will fail automatically by the algorithm.
Next reconciliation should do the job of pool creation.

Reply:
Open to discussion.

Reply:
We will discuss on it
1) Consider a case where We are not able to claim the `Block Device` then above logic will keep on creating stale `BDC` until `SPC` deleted

Reply:
We can clear above stales we will discuss on it